File: WoofAndroid/design-system/src/main/kotlin/com/perrystreet/woof/designsystem/atomic/molecules/button/MolButton.kt
package com.perrystreet.woof.designsystem.atomic.molecules.button

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.minimumInteractiveComponentSize
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.style.TextAlign
import com.perrystreet.woof.designsystem.atomic._tokens.sizing.SizingRoles
import com.perrystreet.woof.designsystem.atomic._tokens.spacing.PaddingRoles
import com.perrystreet.woof.designsystem.atomic.atoms.loading.AtomCircularProgressIndicator
import com.perrystreet.woof.designsystem.atomic.atoms.text.AtomText
import com.perrystreet.woof.designsystem.atomic.atoms.text.roles.TextFontRole
import com.perrystreet.woof.designsystem.atomic.molecules.button.roles.ButtonRole
import com.perrystreet.woof.designsystem.atomic.molecules.button.state.ButtonState
import com.perrystreet.woof.designsystem.theme.Theme

@Composable
fun MolButton(
    text: String,
    onTap: () -> Unit,
    modifier: Modifier = Modifier,
    role: ButtonRole = ButtonRole.Primary,
    state: ButtonState = ButtonState.Enabled,
) {
    val shape = RoundedCornerShape(Theme.radius.m)
    Box(
        modifier = modifier
            .minimumInteractiveComponentSize()
            .defaultMinSize(minHeight = SizingRoles.InteractionHeight.Default.dp)
            .alpha(state.alpha())
            .clip(shape)
            .background(color = role.surfaceColor(), shape = shape)
            .clickable(enabled = state.isClickable, onClick = onTap)
            .semantics(mergeDescendants = true) { this.role = Role.Button }
            .padding(horizontal = PaddingRoles.Element.Expanded.dp),
        contentAlignment = Alignment.Center,
    ) {
        when (state.showsLoadingIndicator) {
            true -> AtomCircularProgressIndicator(
                iconSize = SizingRoles.Icon.M,
                colorRole = role.loadingColorRole,
            )
            false -> AtomText(
                text = text,
                textFontRole = TextFontRole.DisplayH5,
                colorRole = role.textColorRole,
                maxLines = 1,
                textAlign = TextAlign.Center,
            )
        }
    }
}

File: WoofIOS/SwiftPackages/DesignSystem/Sources/DesignSystem/Atomic/Molecules/Button/MolButton.swift
import SwiftUI

public struct MolButton: View {
    @Environment(\.theme) private var theme

    private let text: String
    private let onTap: () -> Void
    private let role: ButtonRole
    private let state: ButtonState

    public init(text: String, onTap: @escaping () -> Void, role: ButtonRole = .primary, state: ButtonState = .enabled) {
        self.text = text
        self.onTap = onTap
        self.role = role
        self.state = state
    }

    public var body: some View {
        Button(action: onTap) {
            ZStack {
                switch state.showsLoadingIndicator {
                case true:
                    AtomCircularProgressIndicator(iconSize: .m, colorRole: role.loadingColorRole)
                case false:
                    AtomText(
                        text: text,
                        textFontRole: .displayH5,
                        colorRole: role.textColorRole,
                        maxLines: 1,
                        textAlign: .center
                    )
                }
            }
            .padding(.horizontal, PaddingRoles.Element.expanded.rawValue)
            .frame(minWidth: SizingRoles.InteractionHeight.default.rawValue, minHeight: SizingRoles.InteractionHeight.default.rawValue)
            .background(role.surfaceColor(from: theme), in: RoundedRectangle(cornerRadius: theme.radius.m))
        }
        .buttonStyle(.plain)
        .disabled(!state.isClickable)
        .opacity(state.alpha(from: theme))
    }
}
